contracts/Router.sol

```diff
@@ -0,0 +1,188 @@
+            
+            address pledger = c.pledgers[i];
+            uint256 amountPledged = c.amountsPledged[pledger];
+
+            // treeToReceive = value pledged * (amountIn / totalPledged)
+            // For example, if 100 DAI is pledged and there's only 50 TREE available
+            // an address that pledged 5 DAI would receive 5 * (50/100) = 2.5 TREE
+            uint256 treeToReceive = amountPledged.mul(amountIn).div(c.totalPledged);
+
+            // Only transfer to EOAs to prevent unexpected reverts if pledge was done using CREATE2
+            // Also, if user ended up unpledging 100%, do not waste a transfer of 0 tokens
+            // note: TREE is already approved to transfer
+            // https://github.com/WhalerDAO/tree-contracts/blob/4525d20def8fce41985f0711e9b742a0f3c0d30b/contracts/TREEReserve.sol#L228
+            if (!Address.isContract(pledger) && amountPledged > 0) {
+                tree.transfer(pledger, treeToReceive);
+                c.treeSold = c.treeSold + treeToReceive;
+            }
+        }
+        numCampaigns++;
+
+        // Return amounts based on https://github.com/WhalerDAO/tree-contracts/blob/4525d20def8fce41985f0711e9b742a0f3c0d30b/contracts/TREEReserve.sol#L217
+        uint256[] memory amounts = new uint256[](2);
+        amounts[0] = c.treeSold;
+        amounts[1] = c.totalPledged;
+
+        emit Rebase(numCampaigns-1, c.treeSold, c.totalPledged);
+    }
+
+
+    function getPledgerId(uint256 _numCampaign, address _addr) private returns (uint256 pledgerId) {
+        Campaign storage c = campaigns[_numCampaign];
+        pledgerId = 0;
+        for (uint i=1; i < c.numPledgers+1; i++) {
+            address pledger = c.pledgers[i];
+            if (pledger == _addr) {
+                pledgerId = i;
+                break;
+            }
+        }
+    }
+
+
+    function setReserveToken(address _newToken) external {
+        require(msg.sender == gov, "UniswapRouter: not gov");
+        reserveToken = IERC20(_newToken);
+        emit SetReserveToken(_newToken);
+    }
+
+
+    function withdrawToken(address _token, address _to, uint256 _amount, bool max) external payable {
+        require(msg.sender == gov, "UniswapRouter: not gov");
+        if (max) {_amount = IERC20(_token).balanceOf(address(this));}
+        IERC20(_token).transfer(_to, _amount);
+        emit WithdrawToken(_token, _to, _amount);
+    }
+
+    function getCampaignTotalPledged(uint256 _numCampaign) public view returns (uint256) {
+        Campaign storage c = campaigns[_numCampaign];
+        return c.totalPledged;
+    }
+
+    function hasPledged(uint256 _numCampaign, address _addr) external view returns (bool pledged) {
+        Campaign storage c = campaigns[_numCampaign];
+        pledged = false;
+        for (i=1;i<c.numPledgers;i++) {
+            if (c.pledgers[i] == _addr && c.amountsPledged[_addr] > 0) {
+                pledged = true;
+                break;
+            }
+        }
```

Comment:
Rather than looping through the `pledgers` mapping, can't we just check
`amountsPledged[_addr] > 0`?

Reply:
Yeah, I may have over-engineered this part